### ISSUE TITLE
Trac wiki migration, improve markup conversion code

### ISSUE THREAD

**Issue by soehms** [sagemath/trac-to-github#18]
A. According to the [sage-devel thread](https://groups.google.com/g/sage-devel/c/xtwaueSUJI0) of the same title, the following tasks must be completed:

1. - [x] Create links with URLs of Trac tickets (to be removed after the migration of the tickets).
   Addressed in PR #15.
2. - [x] Add table alignment instructions.
   Addressed in PR #16.
3. - [x] Sections with fragment names
   Example: `== Legacy sage-trac Account Request == #legacy-account-request` following section [Account Names Mapped to Real Names](https://github.com/sagemath/trac_to_gh/wiki#account-names-mapped-to-real-names) in `Home.md`. Addressed in PR #21.
4. - [x] Various bad links
   Examples: `PageOutline` and `https://groups.google.com/forum/#!forum/sage-packaging` in [Distribution page](https://github.com/sagemath/trac_to_gh/wiki/Distribution). Addressed in PR #19.
5. - [x] Various missing links
   Examples: `[wiki:symbolics/maxima]` and `[https://en.wikipedia.org/wiki/Symbolic_computation]` on [Symbolics page]( https://github.com/sagemath/trac_to_gh/wiki/symbolics)
   Addressed in PR #17
6. - [x] Lines containing nested bullets are not converted.
   Examples: Missing Trac ticket links in nested bullets after `abs_integrate` on [Symbolics page](https://github.com/sagemath/trac_to_gh/wiki/symbolics#integration-tickets). Addressed in PR #23.
7. - [x] Non working queries to the Trac ticket database.
   Examples: `[query:?status=needs_review...` on [Symbolics page](https://github.com/sagemath/trac_to_gh/wiki/symbolics#links-for-these-tickets). Addessed in PR #19.
8. - [x] First lines and headings of pages are omitted.
   Examples: [Distribution page](https://github.com/sagemath/trac_to_gh/wiki/Distribution) (heading is `- Distribution and packaging of SageMath -`). [Infrastructure page](https://github.com/sagemath/trac_to_gh/wiki/Infrastructure) (first line is `This page aims to keep information public about Sage's infrastructure.`)
Addressed in PR #23.
9. - [x] Global references (by URL) to Trac tickets should be localized in order to be converted more easily to their corresponding GitHub issue after ticket migration.
   Examples: Reference to ticket 25206 on [Cygwin64port page](https://github.com/sagemath/trac_to_gh/wiki/Cygwin64port). Addressed in PR #19.

B. Tasks to be performed manually after wiki migration:

1. - [ ] Replace explanations and references to the Trac wiki syntax with explanations and references to the GitHub wiki syntax.
   Examples: [Reporting a bug](https://github.com/sagemath/trac_to_gh/wiki#reporting-a-bug)

C. Tasks to be done after successful ticket migration:

1. - [ ] Remove the link to the Trac ticket so that they point to the corresponding GitHub issue.
2. - [ ] Replace requests to the Trac database with requests to GitHub issues (where possible). But maybe it's better to keep the queries on Trac and add corresponding queries on GitHub.
3. - [ ] Find analogues to Trac macros and insert them at the corresponding hints produced be the script in the new GitHub wiki.

**Comment by kwankyu** [comment]
I guess the final destigation is `sage.wiki`. Who will have the write permission to the wiki? As a wiki, every developer should have the write permission. Is this possible?

**Comment by soehms** [comment]
> Where is the final destination of the migrated wiki? That is, what is the repository that the migrated wiki will be placed?
 
I assume it to be https://github.com/sagemath/sage/wiki.

> ... Perhaps we should do both (fixing manually and improving the script).

Yes, I think so. Improving the script helps also for the ticket migration since surely there are references to wiki pages using all possible syntax variations, as well.

> What is the plan?

From the [migration document](https://github.com/sagemath/sage/wiki/migration-from-trac-to-Git**b#conversion-of-trac-tickets-and-the-trac-wiki-to-github) it is not completely clear. I think, we should consider the Wiki migration as a test case for the ticket migration and thus it should be completed before. For the procedure of the Wiki migration I have the following in mind:

1. Improve the script as long as there are obvious issues that can be fixed easily. Maybe we should create a list of pages that can be completely skipped. For example all pages that explain Trac functionality (having `Trac` or `Wiki` in their names).
2. Run the migration and merge the result into https://github.com/sagemath/trac_to_gh/wiki
3. Let the community check the result.
4. If there are still issues to be fixed generically go back to 1.
5. Do the remaining things manually. At least with respect of item 9 there is still a good deal of manual work left to replace Trac-functionality specific explanations with appropriate links to the GitHub markdown documentation.
6. Let the community check the result again.
7. If there are still issues to be fixed manually go back to 5.
8. Move the result to it's final destination.

After that ticket migration might happen into an appropriate temporary repository at which the community can have a look at it. An iteration similar to 1.-4. can follow.

**Comment by kwankyu** [comment]
Thanks. All you said sounds reasonable. 

Are all the items in A of your first comment "obvious issues that can be fixed easily"? Or any of them is hard to fix but should be implemented because we cannot do it manually?

**Comment by soehms** [comment]
> Are all the items in A of your first comment "obvious issues that can be fixed easily"? Or any of them is hard to fix but should be implemented because we cannot do it manually (that is hard to do manually)?

I think the hardest are done ore addressed, now. What remains are issues caused be some unreasonable lines of code. Deactivating these lines wil fix them. It's just a matter of understanding the purpose of them (see the corresponding remarks in the second comment).

**Comment by kwankyu** [comment]
Okay. I will see if I can help you with some easy items.

**Comment by kwankyu** [comment]
I am working on A.9. You want to convert, for instance, `https://trac.sagemath.org/ticket/25206` to `#25206`. Right?

**Comment by kwankyu** [comment]
I worry that my intervention may make things screwed up. Just let me know when you need a hand on any item.

**Comment by soehms** [comment]
> I am working on A.9. You want to convert, for instance, `https://trac.sagemath.org/ticket/25206` to `#25206`. Right? And these `#...` will be converted to github issues later. Right?

Both answers: Yes!

**Comment by soehms** [comment]
> I worry that my intervention may make things screwed up. Just let me know when you need a hand on any item.

No problem! That can be treated in the PRs. The more people involved the better.

**Comment by soehms** [comment]
All issues known to me are done or addressed, now!

**Comment by kwankyu** [comment]
Great! It looks almost perfect. I looked through all topics, and found the following glitches. Some may need fixing the script and other may need manual edits.

- [] https://github.com/soehms/trac_to_gh/wiki#:~:text=!%5B%5D(ticket_badges.png%2C%20547px)
   trac - https://trac.sagemath.org/attachment/wiki/WikiStart/ticket_badges.png

- [] https://github.com/soehms/trac_to_gh/wiki#:~:text=Displays%20as-,%5C,-%7B%7B%7B%0A%7B%7B%7B%0A%23!python%0Adef%20spam
   trac - https://trac.sagemath.org/#:~:text=name%20diff.-,Markup,-Displays%20as

- [] https://github.com/soehms/trac_to_gh/wiki/GitHub#:~:text=!%5B%5D(https%3A//github.githubassets.com/images/modules/logos_page/Octocat.png%2C%20height%3D200%2C%20right)
   trac - https://github.githubassets.com/images/modules/logos_page/Octocat.png

- []  https://github.com/soehms/trac_to_gh/wiki/GitHub#:~:text=There%20is%20also%20a%20Sage%20mirror%20on%20GitLab.
   trac - https://trac.sagemath.org/wiki/GitHub#:~:text=There%20is%20also%20a%20Sage%20mirror%20on%20GitLab.

- [] https://github.com/soehms/trac_to_gh/wiki/GraphTheoryRoadmap#:~:text=%5BThis%5D(Mathematica%20Graph%20Theory%20)%20page
   trac - https://trac.sagemath.org/wiki/GraphTheoryRoadmap#:~:text=graph%20theory%20software-,This%20page,-aims%20at%20comparing

- [] https://github.com/soehms/trac_to_gh/wiki/Mapple-Graph-Theory-#:~:text=See%20%5BMaple%20Graph%20Theory%20%5D(Maple%20Graph%20Theory%20).
   trac - https://trac.sagemath.org/wiki/Mapple%20Graph%20Theory%20#:~:text=See%20Maple%20Graph%20Theory%20.

- [] https://github.com/soehms/trac_to_gh/wiki/patchbot#:~:text=See%20also%20%5B%2333253%5D(https%3A//trac.sagemath.org/ticket/33253).
   trac - https://trac.sagemath.org/wiki/patchbot#:~:text=multiple%20workflows%20simultaneously.-,See%20also%20%E2%80%8B%2333253.,-Last%20modified%208

- [] many pictures 
 at https://github.com/soehms/trac_to_gh/wiki/patchbot#:~:text=!%5B%5D(https%3A//patchbot.sagemath.org/svg/New%2C%20width%3D48)
   trac - https://patchbot.sagemath.org/svg/New

- [] https://github.com/soehms/trac_to_gh/wiki/PatLeSmithe
   trac - https://trac.sagemath.org/wiki/PatLeSmithe

- [] https://github.com/soehms/trac_to_gh/wiki/SageMath#:~:text=wiki%3Asage.%20The%20full%20(formal)%20name%20of%20the%20software%20that%20is%20nicknamed%20wiki%3Asage.
   trac - https://trac.sagemath.org/wiki/SageMath#:~:text=wiki%3Asage.%20The%20full%20(formal)%20name%20of%20the%20software%20that%20is%20nicknamed%20wiki%3Asage.

- [] https://github.com/soehms/trac_to_gh/wiki/sd51#:~:text=%3Cspan%20style%3D%22color%3A%20blue%22%3EBlue%3C/span%3E%3A%20New
   trac - https://trac.sagemath.org/wiki/sd51#:~:text=Color%20codes%3A-,Blue%3A%20New,-Orange%3A%20Needs%20review

- [] https://github.com/soehms/trac_to_gh/wiki/TicketReports
   trac - https://trac.sagemath.org/wiki/TicketReports

- [] Groupings are gone: ReleaseTours, Sage, combinat, patchbot, sd, Symbolics
   trac - https://trac.sagemath.org/wiki/TitleIndex

**Comment by mkoeppe** [comment]
Thanks a lot for #19, #21, #23! These are great improvements.

**Comment by mkoeppe** [comment]
https://github.com/sagemath/trac_to_gh/wiki/sage-9.6#sage-development-in-the-cloud-with-gitpod has broken inline image

**Comment by mkoeppe** [comment]
> What is the plan?

As soon as we are happy with the quality of the migration of the wiki, we can do a test migration of Trac tickets as well and see if more work on the markup translation is necessary.

Doing the actual final migration and switchover of the Trac wiki is not urgent.

**Comment by kwankyu** [comment]
> Doing the actual final migration and switchover of the Trac wiki is not urgent.

Okay. Would the final migration of wiki occur well before the migration of tickets or at the same time? I think there's no problem in doing wiki migration separately from ticket migration.

**Comment by mkoeppe** [comment]
I agree, the two migrations can be done separately.

**Comment by kwankyu** [comment]
> https://github.com/sagemath/trac_to_gh/wiki/sage-9.6#cylp has two links where the conversion did not work right.

Fixed in trac side.

**Comment by kwankyu** [comment]
> https://github.com/sagemath/trac_to_gh/wiki/sage-9.6#sage-development-in-the-cloud-with-gitpod has broken inline image

Fixed in PR #24.

**Comment by soehms** [comment]
> Thanks a lot for #19, #21, #23! These are great improvements.

Thanks for merging them!

**Comment by soehms** [comment]
> Great! It looks almost perfect. 

Thanks!

> I looked through all topics, and found the following glitches. Some may need fixing the script and other may need manual edits.
> 

Here my diagnostic of these items:

1. , 3. and 8. fail because images with position and size arguments like `[[Image(ticket_badges.png, 547px)]]` are not supported in this Markdown syntax. I doubt that your PR #24 was the right fix for that. I will comment on this later on in the PR.

2. corresponds to my item B1

4. and 10. These are cases of *lazy* syntax omitting `[]` brackets by recognizing keywords like `wiki:` or CamelCase words as links. I think we can do that with the `ConversionHelper->wiki_link` method.

5. I think this is because of the trailing white-space in the page-name. The corresponding page-name on Markdown has a trailing minus: https://github.com/soehms/trac_to_gh/wiki/Mathematica-Graph-Theory-. I'm not sure to better fix this in the Trac-Wiki.

6. It seems to be a redirection for the typo *Mapple* -> *Maple*. The missing link is another instance of 5.

9. and 12. correspond to my item C2.

11. `{{{\n#!html <span style="color: blue">Blue</span>: New\n}}}` should become `$\textcolor{blue}{Blue}$`  (see [this stackoverflow post](https://stackoverflow.com/a/73613791))). This should be possible by the script. I suggest to generally convert the Trac processor `{{{\n#!html...}}}` into flat `html`.

13. The page `TitleIndex` is missing on the GitHub-side since it's author is `trac` which is excluded.

```
        info = source.wiki.getPageInfo(pagename)
        if info['author'] in exclude_authors :
            continue
```

**Comment by kwankyu** [comment]
> Ad 13. The page `TitleIndex` is missing on the GitHub-side since it's author is `trac` which is excluded ...

In E13, I was not talking about the missing `TitleIndex` (which is okay), but about the "groupings" (sections, folders or directories). In trac wiki, the topics can be organized by folder (`symbolic` folder for instance), but in github wiki, there is no such thing. In PR #24, I addressed the problem.

**Comment by soehms** [comment]
> > Ad 13. The page `TitleIndex` is missing on the GitHub-side since it's author is `trac` which is excluded ...
> 
> In E13, I was not talking about the missing `TitleIndex` (which is okay), but about the "groupings" (sections, folders or directories). In trac wiki, the topics can be organized by folder (`symbolic` folder for instance), but in github wiki, there is no such thing. In PR #24, I addressed the problem.

Well, I got that totally wrong! Sorry!

**Comment by kwankyu** [comment]
> Ad 11. `{{{\n#!html <span style="color: blue">Blue</span>: New\n}}}` should become `$\textcolor{blue}{Blue}$` (see [this stackoverflow post](https://stackoverflow.com/a/73613791)). This should be possible by the script. I suggest to generally convert the Trac processor `{{{\n#!html...}}}` into flat `html`.

Thanks. Done.

**Comment by mkoeppe** [comment]
@kwankyu @soehms Many thanks for all your work on this!

**Comment by mkoeppe** [comment]
In https://github.com/sagemath/trac_to_gh/wiki/sage-9.6#cylp I see two more bad links `#33847`, `Meta-ticket #26511`

**Comment by mkoeppe** [comment]
Are images with parameters been handled fully? I see `![](ticket_badges.png, 547px)` in https://github.com/sagemath/trac_to_gh/wiki/sage-9.6#builds-and-checks-of-ticket-branches-on-github-actions

**Comment by kwankyu** [comment]
Did you convert it again with the latest master branch? The images are okay with me. Seeehttps://github.com/kwankyu/trac_to_gh/wiki/WikiStart

**Comment by kwankyu** [comment]
Ah. Your page name is different. It seems to be from the migrated wiki by an old script.

**Comment by mkoeppe** [comment]
Yes, sorry for the noise. I forgot to remove old files and add new files. I'll fix this now

**Comment by mkoeppe** [comment]
I've updated the comments above.

**Comment by kwankyu** [comment]
> Another one in https://github.com/sagemath/trac_to_gh/wiki/SageCodingRoadMap#algorithms-for-generic-codes

Fixed in PR #34

**Comment by mkoeppe** [comment]
In https://github.com/sagemath/trac_to_gh/wiki/Issue-1085 one can see a bad conversion happening within a code block: `This is the Trac macro *...` is inserted

**Comment by kwankyu** [comment]
trac is very slow. Perhaps because of activities here.

**Comment by mkoeppe** [comment]
Another bad conversion (block quotes and code blocks): https://github.com/sagemath/trac_to_gh/blob/main/Issue%2026574.md

**Comment by kwankyu** [comment]
Thanks for the report. 

By the way, the issue number "26574" does not match with the actual trac ticket number "26811" 

https://trac.sagemath.org/ticket/26811

Is this a bug of the script?

**Comment by mkoeppe** [comment]
Yes, that's a bug; I'm working on it

**Comment by mkoeppe** [comment]
In the configuration I used, `filter_issues` is set, so the code that creates placeholder issues is disabled:
```
        if not only_issues and not blacklist_issues and not config.has_option('issues', 'filter_issues') :
```
I'll work on a general solution

**Comment by mkoeppe** [comment]
[Ticket #9757](https://trac.sagemath.org/ticket/9757) is the first missing ticket

**Comment by mkoeppe** [comment]
Code blocks within blockquotes do not work well: https://github.com/sagemath/trac_to_gh/blob/main/Issue%2032652.md

**Comment by mkoeppe** [comment]
Horizontal lines in Trac comments (line with `----`) needs to get an extra newline before it so it does not get interpreted as a section title - https://github.com/sagemath/trac_to_gh/blob/main/Issue%2010037.md

**Comment by mkoeppe** [comment]
Also some quote-reply chains such as "Replying to [comment:8 chapoton]" in https://github.com/sagemath/trac_to_gh/blob/main/Issue%2018823.md are not converted correctly

**Comment by kwankyu** [comment]
> Code blocks within blockquotes do not work well: https://github.com/sagemath/trac_to_gh/blob/main/Issue%2032652.md

With
```
--- a/migrate.cfg.sagetracmigrationarchive
+++ b/migrate.cfg.sagetracmigrationarchive
@@ -49,13 +49,13 @@ migrate: yes
 
 # If defined, import only these issues
 # only_issues: [ 30000, 30001 ]
-#only_issues: [39,]
+only_issues: [32889,]
 
 # If defined, do not import these issues
 # blacklist_issues: [ 268, 843 ]
 
 # If defined, then this is added to the ticket query string to trac
-filter_issues: max=50&order=id&page=1&desc=False
+#filter_issues: max=50&order=id&page=1&desc=False
 #filter_issues: max=2796&order=id&page=2
```
I got 
```
...
SKIP unwanted ticket #32887
SKIP unwanted ticket #32888


## Migrate ticket #32889 (27 changes): Move graphs methods to Polyhed
Create label geometry with color #08517b
# verb='POST' url='https://github.com/sagemath/sagetest/labels' parameters=None headers={'Accept': 'application/vnd.github.symmetra-preview+json'} input={'name': 'geometry', 'color': '08517b'}
# Wrote archive/labels_000001.json
Create label major with color #ff0000
# verb='POST' url='https://github.com/sagemath/sagetest/labels' parameters=None headers={'Accept': 'application/vnd.github.symmetra-preview+json'} input={'name': 'major', 'color': 'ff0000'}
# Wrote archive/labels_000002.json
Create label enhancement with color #008080
# verb='POST' url='https://github.com/sagemath/sagetest/labels' parameters=None headers={'Accept': 'application/vnd.github.symmetra-preview+json'} input={'name': 'enhancement', 'color': '008080'}
# Wrote archive/labels_000003.json
Unmapped users: ['mkoeppe']
Traceback (most recent call last):
  File "/Users/kwankyu/GitHub/trac-to-github/./migrate.py", line 1371, in <module>
    convert_issues(source, dest, only_issues = only_issues, blacklist_issues = blacklist_issues)
  File "/Users/kwankyu/GitHub/trac-to-github/./migrate.py", line 976, in convert_issues
    issue = gh_create_issue(dest, issue_data)
  File "/Users/kwankyu/GitHub/trac-to-github/./migrate.py", line 653, in gh_create_issue
    gh_issue = dest.create_issue(issue_data.pop('title'),
TypeError: Repository.create_issue() got an unexpected keyword argument 'user'
```

**Comment by mkoeppe** [comment]
The script needs a patched version of PyGithub now; use `pip install -r requirements.txt`

**Comment by kwankyu** [comment]
I needed to remove `trac_cache` directory after that.

**Comment by mkoeppe** [comment]
Thanks very much for fixing this! Looking great

**Comment by mkoeppe** [comment]
> the issue number "26574" does not match with the actual trac ticket number "26811"
> 
> https://trac.sagemath.org/ticket/26811
> 
> Is this a bug of the script?

Fixed in f9a937d86cbe890d134791c5be478a9b4feb102e; see for example https://github.com/sagemath/trac_to_gh/blob/main/Issues-26xxx/26811.md

**Comment by kwankyu** [comment]
Thanks. Very nice to look around and check converted issues.

**Comment by kwankyu** [comment]
We may name `md` files with 5 digit number so that they are listed in order.

**Comment by kwankyu** [comment]
> We may name `md` files with 5 digit number so that they are listed in order.

Perhaps not. This inconvenience applies only to issues `00xxx`.

**Comment by mkoeppe** [comment]
Many converted tickets have the text "Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:" but then no commits are listed. What's going on there?
Compare https://trac.sagemath.org/ticket/31580 with https://github.com/sagemath/trac_to_gh/blob/main/Issues-31xxx/31580.md

**Comment by kwankyu** [comment]
I am working on this issue. Almost fixed.

**Comment by mkoeppe** [comment]
Another nested-codeblock problem:
https://github.com/sagemath/trac_to_gh/blob/main/Issues-27xxx/27996.md

**Comment by mkoeppe** [comment]
Also https://github.com/sagemath/trac_to_gh/blob/main/Issues-34xxx/34393.md#comment:4

**Comment by mkoeppe** [comment]
And `@` quoting misbehaves within code-quotes - https://github.com/sagemath/trac_to_gh/blob/main/Issues-34xxx/34393.md#comment:16

**Comment by kwankyu** [comment]
> And `@` quoting misbehaves within code-quotes - https://github.com/sagemath/trac_to_gh/blob/main/Issues-34xxx/34393.md#comment:16

I don't know why `@` is converted to be quoted in the first place. Do you know?

**Comment by mkoeppe** [comment]
I think it's to avoid creating accidental mentions - see also #49

**Comment by mkoeppe** [comment]
problem with a codeblock within a bulleted list
https://github.com/sagemath/trac_to_gh/blob/main/Issues-18xxx/18231.md#comment:3

**Comment by mkoeppe** [comment]
Also https://github.com/sagemath/trac_to_gh/blob/main/Issues-18xxx/18938.md#comment:6

**Comment by mkoeppe** [comment]
Also https://github.com/sagemath/trac_to_gh/blob/main/Issues-27xxx/27870.md#comment:27

**Comment by mkoeppe** [comment]
Bad table conversion: https://github.com/sagemath/trac_to_gh/blob/main/Issues-15xxx/15239.md#comment:2

**Comment by mkoeppe** [comment]
These are all fixed by #56, thanks a lot @kwankyu !
